File: legacy/src/game/pinball-knight/spawn/reaper.ts
/**
 * The REAPER KING — the boss that guards every floor's stairs.
 *
 * Extracted verbatim from core.ts. Its own module because three unrelated
 * places summon it: the floor build, the simulation's exit gate, and the debug
 * panel. The debug verbs used to reach it through `setDebugActionDeps`
 * injection purely because it lived in core; now they can import it.
 */
import { state } from "../state";
import { spawnBoss } from "../boss";
import { showToast } from "../ui";
import { tileCenter } from "../maze/generator";
import { REAPER_HP, REAPER_SCALE, REAPER_SPEED_BASE, REAPER_TINT } from "../constants";
import { isReplica } from "../coop";
import { reaperSheet } from "../render/reaper-sheet";
import { makeZombie } from "../spawn/factory";

/**
 * Spawn the DEATH DEALER: an unkillable blood-red reaper that enters a dozen
 * tiles out from the player (through the walls — it doesn't care) and drifts
 * straight at them, accelerating forever. One per floor; the stairs erase it.
 */
export function spawnReaper(): void {
  const p = state.player;
  if (!p || isReplica()) return; // replica floors get the authority's reaper via snapshot
  state.reaperOut = true;
  const a = Math.random() * Math.PI * 2;
  // Bespoke hooded-and-scythed art (was the ghost sheet dyed with REAPER_TINT).
  const reaper = makeZombie(reaperSheet(), p.x + Math.cos(a) * 12, p.z + Math.sin(a) * 12, REAPER_SPEED_BASE, {
    kind: "reaper",
    hp: REAPER_HP,
  });
  reaper.aggro = true;
  // The sheet is already painted blood-dark, so the tint is now only a faint
  // wash — enough that telegraph/flash clears restore the reaper's colour
  // rather than white, without washing the new art flat.
  reaper.baseTint = REAPER_TINT;
  reaper.sprite.setTint(REAPER_TINT);
  reaper.sprite.mesh.scale.multiplyScalar(REAPER_SCALE);
  state.zombies.push(reaper);
  showToast("☠ THE DEATH DEALER ☠", "it cannot be slain — take the stairs");
  state.shakeT = Math.max(state.shakeT, 0.3);
}

File: crates/pk-core/src/spawn/reaper.rs
//! Reaper King Boss Spawner — The unkillable stair guardian that drifts towards the player.
//!
//! PORTS: `spawn/reaper.ts`

pub const REAPER_DISTANCE_TILES: f32 = 12.0;
pub const REAPER_HP: f64 = 9999.0;
pub const REAPER_SCALE: f64 = 1.25;
pub const REAPER_SPEED_BASE: f64 = 2.2;
pub const REAPER_TINT: u32 = 0x8a1f28;

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct ReaperParams {
    pub hp: f64,
    pub scale: f64,
    pub speed_base: f64,
    pub tint: u32,
    pub aggro: bool,
}

impl ReaperParams {
    pub fn default_params() -> Self {
        Self {
            hp: REAPER_HP,
            scale: REAPER_SCALE,
            speed_base: REAPER_SPEED_BASE,
            tint: REAPER_TINT,
            aggro: true,
        }
    }
}

/// Calculates the spawn position for the Reaper King 12 tiles out from the player at the given angle.
pub fn compute_reaper_spawn_pos(player_x: f32, player_z: f32, angle: f32) -> (f32, f32) {
    (
        player_x + angle.cos() * REAPER_DISTANCE_TILES,
        player_z + angle.sin() * REAPER_DISTANCE_TILES,
    )
}

pub fn spawn_reaper(player_x: f32, player_z: f32, angle: f32) -> (f32, f32, ReaperParams) {
    let (x, z) = compute_reaper_spawn_pos(player_x, player_z, angle);
    (x, z, ReaperParams::default_params())
}
